### DISCUSSION TITLE
How to get a list of edge id from OpenLR code

### DISCUSSION THREAD

**Discussion by Yosef-Irawan** [valhalla/valhalla#3552] [Q&A]
Do we have any function to get a list of edge id from openLR? I think this function will useful. if we want to use traffic data, we will need to get an edge id from openLR code. most traffic providers use openLR in their traffic data.

**Comment by nilsnolde** [comment]
You have to map match to get the IDs. Recently went through that ourselves. It’s not super straight forward if the traffic service came from other geometries than your road network (eg TomTom on OSM) but doable. Devil is in the details;)

**Comment by saschnet** [comment]
Hi Yosef-Irawan,

let me show you how I tried to implement it myself in Python. Since @nilsnolde warns us about the devil in the details (which I have definitely not found yet), he might also give some advice what to do better here:

```python
import httpx
from openlr import LineLocationReference

LEVEL_BITS = 3
TILE_INDEX_BITS = 22
TILE_INDEX_MASK = (2**TILE_INDEX_BITS) - 1
LEVEL_MASK = (2**LEVEL_BITS) - 1

URL = "http://localhost:8002/trace_attributes"


def get_tile_level(id):
  return id & LEVEL_MASK

def get_tile_index(id):
  return (id >> LEVEL_BITS) & TILE_INDEX_MASK

def traffic_tile_id(id):
    return f"{get_tile_level(id)}/{get_tile_index(id)}/0"

def get_shape_from_path(lines:LineLocationReference):
    return [
        {
            "lat": line.lat,
            "lon": line.lon,
        }
        for line in lines.points
    ]

def get_tile_ids_from_driving_path(driving_path:LineLocationReference) -> set[str]:
    shape = get_shape_from_path(driving_path)
    data = {
        "shape": shape,
        "costing":"auto",
        "shape_match":"map_snap",
        "filters":{
            "attributes":
            [
                "edge.id",
            ],
            "action": "include"
        }
    }

    req = httpx.post(URL, json=data, timeout=10)
    req.raise_for_status()
    resp = req.json()

    return {
        traffic_tile_id(edge["id"])
        for edge in resp["edges"]
    }
```

How to use it: Use the function get_tile_ids_from_driving_path by calling it with a Line Reference from the openlr package (see https://openlr-python.readthedocs.io/en/latest/getting-started.html).

The resulting tile id should be as such that you can use the demo at https://github.com/alinmindroc/valhalla_traffic_poc to update the traffic.

**Reply by nilsnolde** [reply]
good start. however, this assumes openlrs (from whichever source) perfectly match to your network, that's rarely going to be the case. even `map_snap` might choose a completely different path than the openlr describes, which you definitely want to avoid. usually you'd involve a more complex flow of `locate`, `route` and only `trace_attributes` once you're fairly sure you have the right path. that'll also involve using bearings, often restricting functional road classes, not using costed routing (i.e. use shortest) etc. also that introduces offsets which you need to account for. that's only for the matching the right edge IDs. then you go on to match the provider's speeds to those edges, break them up where they need breaking up, get all the affected shortcuts and update those as well etc. that's just some of the "details". end-to-end development for a functional, highly performing service updating traffic every minute is quite a beast tbh.

I didn't look too much into the repo you referenced, but one thing is not necessary: re-starting the service when traffic updated. the whole point of using mmap'd tars is so you can share the state between the processes (valhalla and traffic-updater). re-starting the service will destroy highly valuable caches. mmaps is smth I also didn't know too well before getting involved with valhalla.

**Reply by Yosef-Irawan** [reply]
Hi @nilsnolde, can you share a bit about the traffic update process? is it have some Valhalla commands that need to run after updating the traffic? I have updated the traffic while Valhalla running but the response isn't showing the latest traffic data.

**Comment by kevinkreiser** [comment]
I owe others an explanation of openlr matching so I'll post back here with a synopsis in a little bit

**Reply by geoffreyp** [reply]
I am interested in learning more about this topic! Traffic management seems to be integrated in valhalla, but it is very difficult as there is no concrete example and documentation on the subject.

**Reply by schierkolk** [reply]
I am interested as well.
